### PR TITLE
deps: Take all minor upgrades; ~compileSdkVersion 33;~ camera-roll 5.x

### PR DESCRIPTION
**Pull request by gnprice**
The final commit here runs `yarn upgrade`. The others represent changes that were needed in order to unblock that.

Specifically, the new react-native-image-picker version requires `compileSdkVersion` to be bumped to 33, the newest: https://github.com/react-native-image-picker/react-native-image-picker/issues/2018#issuecomment-1307887099

That in turn requires upgrading react-native-gesture-handler beyond the `~` constraint we'd had on it, and requires upgrading react-native-camera-roll from 4.x to 5.0.2:
  https://github.com/react-native-cameraroll/react-native-cameraroll/issues/429

@chrisbobbe would you do some manual testing of this PR on iOS? I've done so on Android… but the app's iOS build seems to be broken on my Mac, and I've spent a couple of hours on that so far today without success. I'll start a separate conversation (edit: [here](https://chat.zulip.org/#narrow/stream/48-mobile/topic/stale.20FBReactNativeSpec.20causes.20iOS.20build.20failure)) on that, but in the meantime figure it's better to unblock this.

### CODE REVIEW

**Comment by gnprice** [comment]
(Bump, now that I've properly filled in the title and description, for the sake of causing email notifications with a better subject line.)

**Review by chrisbobbe — COMMENTED** [review]
Thanks!

I've just tested this on iOS (my iPhone 13 Pro, running iOS 16.1).

The `react-native-image-picker` upgrade seems problematic, unfortunately. Here's some background:

- For a while, we've had a problem (tracked in the stalled PR #4475) where, for image uploads on iOS, the app sends a copy of the original image with a much-increased size.
- ISTR at one point we diagnosed the problem as follows: Apple's API for making a GIF copy (`UIImageJPEGRepresentation`) gives a result with a much-increased size when you ask for quality 1.0. The image was being passed through this API, with quality 1.0, _twice_: once in `react-native-image-picker` and once in `react-native` internals.
- We've hoped that https://github.com/facebook/react-native/commit/f78526ce3, released in RN v0.66, would relieve the symptom by removing the `UIImageJPEGRepresentation` call in RN. We haven't tested that hypothesis; I hope it removed the call but I don't know yet.
- I've been vaguely optimistic that the burn-it-down rewrite of `react-native-image-picker` in v3 might've completed the fix by removing _its_ call. I think that hasn't happened? 

The `react-native-image-picker` upgrade in your PR gives us https://github.com/react-native-image-picker/react-native-image-picker/pull/2006, "fix: the fileSize of the selected image is bigger than the origin one in ios", released in [4.8.5](https://github.com/react-native-image-picker/react-native-image-picker/releases/tag/v4.8.5).

But our app's problem hasn't gone away. At the tip of this PR branch, having run `yarn`,
- I tested the upload-from-library flow. The image was taken last Friday by my iPhone, and its detail page in my Photos app claims it's 3.5MB. When it finished uploading, I downloaded it from the Zulip web-app lightbox, and it's showing as 8.1MB in Finder.
- I tested the upload-from-camera flow. The image was saved to my iPhone; its detail page in my Photos app claims it's 1.4MB. When it finished uploading, I downloaded it from the Zulip web-app lightbox, and it's showing as 3.7MB in Finder. Also—and this is consistent with https://github.com/react-native-image-picker/react-native-image-picker/issues/2022, which blames the 4.8.5 upgrade—the image was rotated 90° counterclockwise.

I didn't test on `main` to compare, but the 90°-rotation bug seems new and something we'd want to avoid.

Here are things I did that succeeded. (I used GitHub's "rich diff" feature for viewing all the `yarn.lock` changes from this PR, and went down the list. Helpful UI! 🙂)

- I successfully built and ran the app
- The upgraded `@react-native-camera-roll/camera-roll` still seems to work; I successfully downloaded an image from the lightbox
- The upgraded `react-native-gesture-handler` still seems to work (at least when not in "Debug with Chrome" mode; gestures are known not to work in that mode):
  - Swipe-right-from-left to go back
  - Swipe between top tabs in the see-who's-reacted screen
- The upgraded `@expo/react-native-action-sheet` still seems to work; I was able to open the action sheet and view read receipts for a message.
- The upgraded `expo-apple-authentication` still seems to work; I successfully signed in with Apple.
- The upgraded `react-native-safe-area-context` still seems to work; I didn't notice new bugs with the safe-area insets, even when I rotated to landscape mode
- I didn't see any new bugs that would point to something wrong in `react-native-screens`, but let's keep an eye out because the [changelog](https://github.com/software-mansion/react-native-screens/releases) lists a lot of stuff between 3.13.1 and 3.18.2.
- The upgraded `react-native-vector-icons` still seems to work; didn't notice any wrong-looking icons.
- The upgraded `react-native-webview` still seems to work; didn't notice anything wrong in the webview.
- Changes to `react-redux`, `redux-thunk`, and `reselect` seem small from their changelogs, and I didn't see anything wrong with these while testing all the other stuff.


Things I didn't test:
- Sentry; I hope we'd notice anything gone wrong in the next beta release

Other things I happened to notice:

- If we were using `tsflower` for `expo-sqlite`, we'd automatically get a type adjustment that would support using `null` as a query argument: https://github.com/expo/expo/pull/18078

File: android/build.gradle [hunk 1]
```diff
@@ -19,7 +19,7 @@ buildscript {
         //   https://medium.com/androiddevelopers/picking-your-compilesdkversion-minsdkversion-targetsdkversion-a098a0341ebd
         // What's the latest?  Consult this list:
         //   https://developer.android.com/studio/releases/platforms
-        compileSdkVersion = 31
+        compileSdkVersion = 33
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply android/build.gradle:22]: Hrm, I guess this is in conflict with my finding in https://github.com/zulip/zulip-mobile/pull/5543#pullrequestreview-1174649778 that 4.8.5 is problematic. 😵

**Comment by gnprice** [comment]
> But our app's problem hasn't gone away. At the tip of this PR branch, having run `yarn`,
> 
> * I tested the upload-from-library flow. The image was taken last Friday by my iPhone, and its detail page in my Photos app claims it's 3.5MB. When it finished uploading, I downloaded it from the Zulip web-app lightbox, and it's showing as 8.1MB in Finder.
> * I tested the upload-from-camera flow. The image was saved to my iPhone; its detail page in my Photos app claims it's 1.4MB. When it finished uploading, I downloaded it from the Zulip web-app lightbox, and it's showing as 3.7MB in Finder. Also—and this is consistent with [react-native-image-picker/react-native-image-picker#2022](https://github.com/react-native-image-picker/react-native-image-picker/issues/2022), which blames the 4.8.5 upgrade—the image was rotated 90° counterclockwise.
> 
> I didn't test on `main` to compare, but the 90°-rotation bug seems new and something we'd want to avoid.

Ugh. Fun. Thanks for finding this.

There's that RN issue that also caused this sort of bloat. It looks like the commit you mentioned was meant to fix this issue:
https://github.com/facebook/react-native/issues/27099
and that issue reportedly wasn't actually fixed in v0.66:
https://github.com/facebook/react-native/issues/27099#issuecomment-1006748552
So that could be enough to explain the file sizes you saw, right?

---

For the rotation: someone kindly sent a PR to fix it:
https://github.com/react-native-image-picker/react-native-image-picker/pull/2036
and it looks pretty reasonable. So one possible solution would be to run the PR branch. (It's already on top of the latest release, so no need even to rebase.)

**Comment by gnprice** [comment]
Thanks for the review and the thorough testing! Replied on the react-native-image-picker issue just above; I'll look at your other comments tomorrow, and perhaps see if I can produce a version that holds that back in order to unblock the rest.

File: types/@react-native-camera-roll/camera-roll/lib/typescript/CameraRoll.js.flow [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    Photos: string,
+    ...
+  };
+  static deletePhotos(photoUris: Array<string>): void;
+  static save(tag: string, options?: SaveToCameraRollOptions): Promise<string>;
+  static saveToCameraRoll(tag: string, type?: 'photo' | 'video' | 'auto'): Promise<string>;
+  static getAlbums(params?: GetAlbumsParams): Promise<Album[]>;
+  static getParamsWithDefaults(params: GetPhotosParams): GetPhotosParams;
+  static getPhotos(params: GetPhotosParams): Promise<PhotoIdentifiersPage>;
+  static iosGetImageDataById(internalID: string, convertHeicImages?: boolean): Promise<PhotoIdentifier>;
```
[the remark this answers is not in the document]
Comment (gnprice) [reply types/@react-native-camera-roll/camera-roll/lib/typescript/CameraRoll.js.flow:106]: Hmm, indeed. Puzzling that it passed CI, then.

Ah, I see. It's that in the `tools/test --all` case, the `prettier` suite relies on this line:
```
    patterns=( "${@/%\///**/*.js}" ) # replace trailing `/` with `/**/*.js`
```
And that line wasn't adapted to reflect our wanting to cover `.js.flow` files as well as `.js` files.

I'll see about fixing that too, perhaps as a separate PR.

File: types/patches/0043-tsflower-r-n-safe-area-c-Substitute-for-InstanceType.patch [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+diff --git types/react-native-safe-area-context/lib/typescript/SafeAreaView.js.flow types/react-native-safe-area-context/lib/typescript/SafeAreaView.js.flow
+index 679c3bcbe..a25801356 100644
+--- types/react-native-safe-area-context/lib/typescript/SafeAreaView.js.flow
++++ types/react-native-safe-area-context/lib/typescript/SafeAreaView.js.flow
+@@ -9,7 +9,7 @@ import type {
+ import * as React from 'react';
+ import { type NativeSafeAreaViewProps } from './SafeArea.types';
+ import NativeSafeAreaView from './specs/NativeSafeAreaView';
+-type NativeSafeAreaViewInstance = InstanceType<typeof NativeSafeAreaView>;
++type NativeSafeAreaViewInstance = $Call<typeof NativeSafeAreaView>;
```
[the remark this answers is not in the document]
Comment (gnprice) [reply types/patches/0043-tsflower-r-n-safe-area-c-Substitute-for-InstanceType.patch:20]: Hmm. Yeah, on second look `$Call` isn't right at all. That was an attempt to translate TS's `InstanceType` -- but `InstanceType` takes a *constructor* to the type it returns *called as a constructor*, with `new`, which isn't the same as the normal function call that `$Call` represents.

Unpacking the definition of `NativeSafeAreaView`, its type is a `React.AbstractComponent<…>`. So yeah, I think `React.ElementRef` is appropriate.

File: android/build.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (gnprice) [reply android/build.gradle:22]: > At this commit, `tools/test native` failed for me with this:

Hmm, good catch, thanks. I thought I'd checked that sort of thing, but apparently missed this.


> That error looks like it matches [react-native-image-picker/react-native-image-picker#2013](https://github.com/react-native-image-picker/react-native-image-picker/issues/2013), which links to PR [react-native-image-picker/react-native-image-picker#2003](https://github.com/react-native-image-picker/react-native-image-picker/pull/2003), which links to a PR [react-native-image-picker/react-native-image-picker#1955](https://github.com/react-native-image-picker/react-native-image-picker/pull/1955) that was merged and reportedly released in `react-native-image-picker` 4.9.0.

Ugh. Looks like they fixed the error making it *possible* to build on the new SDK:
https://github.com/react-native-image-picker/react-native-image-picker/pull/1955/files#diff-68f6dace32eaf1f8c9e342c552cb4984338d3e185ccb1552d72fbb5df278f77e
in the same commit where they also added that feature that made it *necessary* to do so. Bad choice.

---

Well, I think this was also the one thing that made it necessary to bump the compileSdkVersion. So I may be able to get a working upgrade branch by:
* holding back react-native-image-picker to 4.8.4, before both this change and the preceding one that introduced that rotation bug;
* and holding back compileSdkVersion at 31.

I'll go try that now. Then if that works, straightening things out on react-native-image-picker and doing the compileSdkVersion bump can be a follow-up.

**Comment by gnprice** [comment]
OK, updated!

I held back react-native-image-picker at 4.8.4 (our current version, and the last one before that rotation bug), and then held back compileSdkVersion at 31 as a result, as per this subthread: https://github.com/zulip/zulip-mobile/pull/5543#discussion_r1019564281

Also made smaller changes discussed above.

Marked as non-draft thanks to your testing.

File: types/@react-native-camera-roll/camera-roll/lib/typescript/CameraRoll.js.flow (hunk 2 above)
[the remark this answers is not in the document]
Comment (gnprice) [reply types/@react-native-camera-roll/camera-roll/lib/typescript/CameraRoll.js.flow]: > I'll see about fixing that too, perhaps as a separate PR.

Filed #5546 for this, and sent #5547.

**Comment by chrisbobbe** [comment]
Thanks, LGTM! Merged.

File: android/build.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply android/build.gradle]: > straightening things out on react-native-image-picker and doing the compileSdkVersion bump can be a follow-up.

#5618

**Pull request merged**